docs/ai/conceptual/understanding-openai-functions.md

```diff
@@ -0,0 +1,66 @@
+---
+title: "Understanding OpenAI Function Calling"
+description: "Understand how function calling enables you to integrate external tools with your OpenAI application."
+author: haywoodsloan
+ms.topic: concept-article 
+ms.date: 04/26/2024
+
+#customer intent: As a .NET developer, I want to understand OpenAI function calling so that I can integrate external tools with AI completions in my .NET project.
+
+---
+
+# Understand OpenAI function calling
+
+Function calling is a feature of OpenAI models that supports describing functions and their arguments in a JSON format when making an completion request. The model won't invoke the function itself, but will return a JSON output describing what functions should be called and their arguments.
+
+Function calling simplifies connecting external tools to your AI model. Each tool is specified to the model as one or more functions. The models will then decide if a function should be used based on the prompt question. The model will use the results of the function calls to build a more accurate and consistent response.
+
+Potential use cases for function calling include:
+
+* Answering questions by calling external APIs, e.g. sending emails or getting the weather.
+* Answering questions with info from an internal datastore, e.g. aggregating sales data to answer, "What are my best selling products?".
+* Create structured data from text info, e.g. building a user info object with details from the chat history.
+
+## Calling functions with OpenAI
+
+The general steps for calling functions with an OpenAI model are:
+
+* Send the user's question as request with functions defined in the [`tools` parameters](https://platform.openai.com/docs/api-reference/chat/create#chat-create-tools)
+* The model decides which functions, if any, to call. The output will contain a JSON object listing the function calls and their arguments.
+    > [!NOTE]
+    > The model may hallucinate additional arguments
+* Parse the output and call the requested functions with their specified arguments.
+* Send another request with the function results included as a new message.
+* The model responds with more function call requests or an answer to the user's question.
+  * Continue invoking the requested function calls until the model responds with an answer.
+
+You can force the model to request a specific function by setting the `tool_choice` parameter to the function's name. You can also force the model to response with a message for the user by setting the `tool_choice` parameter to `"none"`.
+
+## Calling functions in parallel
+
+Parallel function calling enables a model to request multiple function calls in one output. The results of each function call are included together in one response back to the model. This reduced the number of API requests and time needed to generate an answer. Each function result is included as a new message in the conversation with a `tool_call_id` matching the `id` of the function call request.
```

Comment:
Is this a setting or something you do with code?

Reply:
This is just a feature of some models. If a model supports it you'll get multiple function calls requested in one response from the model. If not you'll just go back and forth with the model as it requests one function call at a time. Devs don't specify anything special, it's just a difference in the behavior of some models. For the Semantic Kernel for example, the Kernel handles orchastrating the calls in the first place so devs won't even see if each function call was it's own response or if they were in parallel.

Reply:
I reworded to indicated only some models suppor this.